### DISCUSSION TITLE
Find specified object in a dynamically linked library

### DISCUSSION THREAD

**Discussion by hhy2022** [idaholab/moose#20787] [Q&A General] [answered]
Hello, I was trying to develop one simple advection application and I followed the guidance on this page: [https://mooseframework.inl.gov/source/kernels/ConservativeAdvection.html]

I got this message:
```
*** ERROR ***
A 'InflowBC' is not a registered object.

If you are trying to find this object in a dynamically linked library, make sure that
the library can be found either in your "Problem/library_path" parameter or in the
MOOSE_LIBRARY_PATH environment variable.
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 1
Finished: Exit code: 1
```

I understood that I didn't link my appliction to the right library and I tried to do so by enabling some physical module in the "Makefile" file and re-compile the application.
After that, I check my library with the comman `ls src/* include/*` and the messages were:
src/main.C
src/main.x86_64-conda-linux-gnu.opt.lo
src/main.x86_64-conda-linux-gnu.opt.lo.d

include/base:
CatApp.h

src/base:
CatApp.C                              CatApp.x86_64-conda-linux-gnu.opt.lo.d
CatApp.x86_64-conda-linux-gnu.opt.lo

It seemed that I didn't change anything. I might use the wrong way to link my application to the library and I am confused on how to do so.
Besides, If I only know the specified parts (like ConservativeAdvection Kernel) I would like to use, which physical module should I enable to make sure I could use those parts successefully? I checked the link [https://mooseframework.inl.gov/modules/](url) and didn't find the ConservativeAdvection Kernel at any module, so I am wondering if there is  a list on the module detailed information.

Thank you so much for your help.

**Comment by GiudGiud** [comment, marked as answer]
Hello

linking to an application is not going to modify the source and headers in your application. It will just link to something that was compiled elsewhere.
Where is this InflowBC you need? I dont see it in the moose modules. 
ConservativeAdvection is a framework object, so it's available by default

Guillaume

**Reply by hhy2022** [reply]
InflowBC was used in the boundary conditions. Since it was a advection problem, I would need to defined mass flow inlet and outlet.
I followed the guidance on this page [](https://mooseframework.inl.gov/source/kernels/ConservativeAdvection.html). 
So this was a test kernel and if I would like to use it, I had to copied some of the .h and .c files to my own directory and compiled them. Right?

**Reply by hhy2022** [reply]
Sorry, I link was: [https://mooseframework.inl.gov/source/kernels/ConservativeAdvection.html](url)

**Reply by GiudGiud** [reply]
ConservativeAdvection is not a test kernel. You can use it in your app by default because it's part of the framework kernels. 

InflowBC is a test object in MOOSE. So yes you would need to move it to your application

**Reply by hhy2022** [reply]
Got you. Thank you so much for your help.
